### ISSUE TITLE
Misleading Exception

### ISSUE THREAD

**Issue by njw6622** [javascript-obfuscator/javascript-obfuscator#760]
When I want to **obfuscate one directory** like this:
`javascript-obfuscator <srcDir> --output <destDir> --config <myConfig>`
If in my config I assign value to `sourceMapFileName` , then an exception thown:
```
----- now run cmd: 'javascript-obfuscator dist\dev-bundle       --output dist\prod-bundle       --config buildScripts\obfuscate.config.json' -----
Error: EPERM: operation not permitted, mkdir 'D:\'
    at Object.mkdirSync (fs.js:947:3)
    at mkdirpNativeSync (D:\workspace\HiEarthSdk\node_modules\mkdirp\lib\mkdirp-native.js:25:17)
    at Object.sync (D:\workspace\HiEarthSdk\node_modules\mkdirp\index.js:21:7)
    at t.ObfuscatedCodeWriter.writeFile (D:\workspace\HiEarthSdk\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\utils\ObfuscatedCodeWriter.ts:82:16)
    at j.apply (D:\workspace\HiEarthSdk\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\JavaScriptObfuscatorCLI.ts:195:39)
    at j.processSourceCodeWithSourceMap (D:\workspace\HiEarthSdk\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\decorators\Initializable.ts:61:39)
    at j.apply (D:\workspace\HiEarthSdk\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\JavaScriptObfuscatorCLI.ts:178:18)
    at j.processSourceCode (D:\workspace\HiEarthSdk\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\decorators\Initializable.ts:61:39)
    at D:\workspace\HiEarthSdk\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\JavaScriptObfuscatorCLI.ts:165:22
    at Array.forEach (<anonymous>)
(node:6204) UnhandledPromiseRejectionWarning: Error: Command failed: javascript-obfuscator dist\dev-bundle       --output dist\prod-bundle       --config buildScripts\obfuscate.config.jso
n

```

The exception is too terrible and I found out reason after one day...

## Expected Behavior

So when obfuscating one directory the option `sourceMapFileName` should not be set?
If set I want get correct prompt... Thanks!

## Your Environment

* Obfuscator version used: windows10
* Node version used: v14.10.0

**Comment by sanex3339** [comment]
https://stackoverflow.com/questions/34600932/npm-eperm-operation-not-permitted-on-windows

**Comment by njw6622** [comment]
No, I think the root reason is **OUTPUT PATH IS WRONG**
You can:
1. set any value to option `sourceMapFileName` in the config file
2. obfuscate a **directory**: `javascript-obfuscator <srcDir> --output <distDir> --config <yourConfigFile>`

So I can get the exception say `Error: EPERM: operation not permitted, mkdir 'D:\'`
Of course you can't make a dir named "D:\\" but the app want to 

#  Reproduce Repos

I create a reproduce repos: https://github.com/njw6622/test-obfuscator.git
You can clone it for test
test cmds:
```shell
git clone https://github.com/njw6622/test-obfuscator.git
cd test-obfuscator
npm i
npm test
```

**Comment by sanex3339** [comment]
Ok. But the prblem is that i can't test it, because i'm developing under UNIX environment (

**Comment by da411d** [comment]
Yes, i managed to reproduce the problem.
Will investigate tonight.

**Comment by sanex3339** [comment]
Thank you.
The main problem, i'm not sure how to properly test cross-platform code.

**Comment by da411d** [comment]
On Windows we have WSL. Maybe on UNIX you have something similar?

**Comment by da411d** [comment]
First, I added some `console.log` to `mkdirp` and `javascript-obfuscator`.
On the left is displayed current configuration, on the right - test proccess.
We can see that this problem is not caused by saving obfuscated file, this is problem with saving source map file.
![image](https://user-images.githubusercontent.com/12462454/94313724-ea9c4880-ff87-11ea-8e43-ab251df979f0.png)
<details>
<summary>Code from screenshot</summary>
<b>config.json</b>
<pre>
{
    "optionsPreset": "default",
    "sourceMap": true,
    "sourceMapFileName": "abcdefg"
}
</pre>

<b>Console output</b>
<pre>
writeFile: "dist\src\a.js" obfuscatedCode.getObfuscatedCode()
mkdirpSync: dist\src undefined
writeFile: "/abcdefg.js.map" obfuscatedCode.getSourceMap()
mkdirpSync: / undefined

[javascript-obfuscator-cli] Error in file: src\a.js...
Error: EPERM: operation not permitted, mkdir 'D:\'
    at Object.mkdirSync (fs.js:840:3)
    at mkdirpNativeSync (D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\mkdirp\lib\mkdirp-native.js:26:17)
    at Object.this (D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\mkdirp\index.js:22:7)
    at t.ObfuscatedCodeWriter.ObfuscatedCodeWriter [as writeFile] (D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\utils\SourceCodeReader.ts:21:32)
    at F.apply (D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\JavaScriptObfuscatorCLI.ts:226:74)
    at F.obfuscatedCodeWriter (D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\decorators\Initializable.ts:61:39)
    at F.apply (D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\JavaScriptObfuscatorCLI.ts:197:14)
    at F.value (D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\decorators\Initializable.ts:61:39)
    at D:\_PROJECTS\_WEB\test2\test-obfuscator\node_modules\javascript-obfuscator\dist\webpack:\JavaScriptObfuscator\src\cli\JavaScriptObfuscatorCLI.ts:188:5
    at Array.forEach (<anonymous>)
npm ERR! Test failed.  See above for more details.
</pre>
</details>


The exception is caused by `mkdirp` when it attempts to create folder "/".
![image](https://user-images.githubusercontent.com/12462454/94311274-a444ea80-ff83-11ea-8e33-e265afe31129.png)
<details>
<summary>Code from screenshot</summary>
<b>test.js</b>
<pre>
const mkdirp = require("mkdirp");
mkdirp.sync("/")
</pre>

<b>Console output</b>
<pre>
internal/fs/utils.js:230
    throw err;
    ^

Error: EPERM: operation not permitted, mkdir 'D:\'
[90m    at Object.mkdirSync (fs.js:840:3)[39m
    at mkdirpNativeSync (D:\_PROJECTS\_WEB\test2\test\node_modules\[4mmkdirp[24m\lib\mkdirp-native.js:25:17)
    at Function.mkdirpSync [as sync] (D:\_PROJECTS\_WEB\test2\test\node_modules\[4mmkdirp[24m\index.js:21:7)
    at Object.<anonymous> (D:\_PROJECTS\_WEB\test2\test\test.js:2:8)
[90m    at Module._compile (internal/modules/cjs/loader.js:1158:30)[39m
[90m    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)[39m
[90m    at Module.load (internal/modules/cjs/loader.js:1002:32)[39m
[90m    at Function.Module._load (internal/modules/cjs/loader.js:901:14)[39m
[90m    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:74:12)[39m
[90m    at internal/main/run_main_module.js:18:47[39m {
  errno: [33m-4048[39m,
  syscall: [32m'mkdir'[39m,
  code: [32m'EPERM'[39m,
  path: [32m'D:\\'[39m
}
</pre>
</details>



The problem is that function [getOutputSourceMapPath](https://github.com/javascript-obfuscator/javascript-obfuscator/blob/master/src/cli/utils/ObfuscatedCodeWriter.ts#L81) is generating strange paths.
I copied function to separate file and did some random tests.
First example is for our case. The second and last also look the most strange for me.
![image](https://user-images.githubusercontent.com/12462454/94314383-0eac5980-ff89-11ea-9802-653402b7e00d.png)
<details>
<summary>Code from screenshot</summary>
<b>test_codeWriter__getOutputSourceMapPath.js</b>
<pre>
const getOutputSourceMapPath = (outputCodePath, sourceMapFileName = '') => {
  if (sourceMapFileName) {
    outputCodePath = `${outputCodePath.substring(
        0, outputCodePath.lastIndexOf('/')
    )}/${sourceMapFileName}`;
  }

  if (!/\.js\.map$/.test(outputCodePath)) {
    outputCodePath = `${outputCodePath.split(".")[0]}.js.map`;
  } else if (/\.js$/.test(outputCodePath)) {
    outputCodePath += '.map';
  }

  return outputCodePath;
}

const tests = [
  ["", "abcdefg"],
  ["./", "./abcdefg"],
  ["/dir/dist", "abcdefg"],
  ["/dir", "abcdefg"],
  ["dir/dist", "abcdefg"],
  ["dir", "abcdefg"],
  
  ["D:/dir/dist", "abcdefg"],
  ["D:/dir/dist", "../abcdefg"],
];

tests.forEach(args => {
  const result = getOutputSourceMapPath(...args);
  console.log(`\ngetOutputSourceMapPath(${args.map(a => `"${a}"`).join(", ")}) = ${result}`);
});
</pre>

<b>Console output</b>
<pre>
getOutputSourceMapPath("", "abcdefg") = /abcdefg.js.map

getOutputSourceMapPath("./", "./abcdefg") = .js.map

getOutputSourceMapPath("/dir/dist", "abcdefg") = /dir/abcdefg.js.map

getOutputSourceMapPath("/dir", "abcdefg") = /abcdefg.js.map

getOutputSourceMapPath("dir/dist", "abcdefg") = dir/abcdefg.js.map

getOutputSourceMapPath("dir", "abcdefg") = /abcdefg.js.map

getOutputSourceMapPath("D:/dir/dist", "abcdefg") = D:/dir/abcdefg.js.map

getOutputSourceMapPath("D:/dir/dist", "../abcdefg") = D:/dir/.js.map
</pre>
</details>

---

Also i remember that [getOutputCodePath](https://github.com/javascript-obfuscator/javascript-obfuscator/blob/master/src/cli/utils/ObfuscatedCodeWriter.ts#L38) also had problems on Windows [#576](https://github.com/javascript-obfuscator/javascript-obfuscator/issues/576#issuecomment-652684639)

So I think that working with paths needs refactoring and more tests, but I don't know good ways to do it. Maybe [path](https://nodejs.org/api/path.html) has something acceptable? At least, it may exclude bugs with splitting by ".".

**Comment by hecht-software** [comment]
Same here on Windows even with one file instead of whole directory. My workaround is to specify a full path for the source map file. I guess this doesn't work for directories.

I remember that I had a problem creating nested directories on Windows and Linux a few years ago. I didn't look into the problem in detail, but I leave you this code which is still in use and maybe helpul here if it's the same issue. It works like `mkdir -p` but takes Windows traits into account:

```js
icl.pathDelimRegex = /[\/\\]/;

icl.getDirPathByFilePath = function(path)
{
   var pathEntities = path.split(icl.pathDelimRegex);
   pathEntities.pop();
   return pathEntities.join("/");
};

icl.createDirDone = function(dir, fprops)
{
   if (fprops === undefined)
   {
      fprops = 0o777;
   }

   var d = icl.Deferred();

   function createDirRcsv(dirEntities, i)
   {
      var curDir = dirEntities.slice(0, i).join("/");
      fs.mkdir(curDir, fprops, function(err) {
         if (err && err.code !== "EEXIST")
         {
            d.reject(err);
         }
         else if (i <= dirEntities.length)
         {
            createDirRcsv(dirEntities, i+1);
         }
         else
         {
            d.resolve();
         }
      });
   }

   var dirEntities = dir.split(icl.pathDelimRegex);
   var i = 1;
   if (/^[\\\/]{2}/.test(dir))
   {
      i = 3;
   }
   else if (!dirEntities[0].length || dirEntities[0][dirEntities[0].length - 1] === ":")
   {
      // absolute path beginning with "/" (Linux etc.) or "X:" (Windows)
      i = 2;
   }
   createDirRcsv(dirEntities, i);

   return d.promise();
};

// Then it's used like that:
icl.createDirDone(icl.getDirPathByFilePath(path)).then(/*...*/);
```

**Comment by sanex3339** [comment]
I managed to setup TravisCI build under windows.
https://travis-ci.com/github/javascript-obfuscator/javascript-obfuscator/jobs/395863667

It fails with the same error. So, now i'll try to fix it

**Comment by sanex3339** [comment]
Anybody can test this PR:
https://github.com/javascript-obfuscator/javascript-obfuscator/pull/772 ?

**Comment by sanex3339** [comment]
Merged to the `master` branch. I'll release a new version soon.

**Comment by fishbone1** [comment]
Thank you!
I couldn't test it yesterday. Did someone else test it?

**Comment by sanex3339** [comment]
You can test it right now. I'll plan to release the next version only in 10 hours.

**Comment by sanex3339** [comment]
Anybody tested this?

**Comment by fishbone1** [comment]
Sorry, I need more time, I think I'm ready in a few hours

**Comment by fishbone1** [comment]
I tested it and it works with master. I also reproduced the error from last week with the old version before, to make sure I can trust the positive result. Thanks for fixing it!

**Comment by sanex3339** [comment]
Nice! I'll release the new version soon.

**Comment by sanex3339** [comment]
Released as `2.5.0`
